internal/provider/provider.go

```diff
@@ -196,50 +240,147 @@ func (p *pingfederateProvider) Configure(ctx context.Context, req provider.Confi
+				)
+			}
+		}
+
+		configClientSecret, ok := config.OAuth.Attributes()["client_secret"].(types.String)
+		if ok {
+			clientSecret = configClientSecret.ValueString()
+		} else {
+			clientSecret = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_CLIENT_SECRET")
+			if clientSecret == "" {
+				resp.Diagnostics.AddError(
+					"Unable to find client_secret",
+					"client_secret cannot be an empty string. Either set it in the configuration or use the PINGFEDERATE_PROVIDER_OAUTH_CLIENT_SECRET environment variable.",
+				)
+			}
+		}
+
+		configScopes, ok := config.OAuth.Attributes()["scopes"].(types.List)
+		if ok {
+			configScopes.ElementsAs(ctx, &scopes, false)
+		} else {
+			scopes = []string{os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_SCOPES")}
+		}
+
+		configTokenUrl, ok := config.OAuth.Attributes()["token_url"].(types.String)
+		if ok {
+			tokenUrl = configTokenUrl.ValueString()
+		} else {
+			tokenUrl = os.Getenv("PINGFEDERATE_PROVIDER_OAUTH_TOKEN_URL")
+			if tokenUrl == "" {
+				resp.Diagnostics.AddError(
+					"Unable to find token_url",
+					"token_url cannot be an empty string. Either set it in the configuration or use the PINGFEDERATE_PROVIDER_OAUTH_TOKEN_URL environment variable.",
+				)
+			}
+		}
+
+		if clientId != "" && clientSecret != "" && tokenUrl != "" {
+			hasOauthConfig = true
+		} else {
+			hasOauthConfig = false
 		}
 	}
 
-	// User must provide a password to the provider
-	var password string
-	if config.Password.IsUnknown() {
-		// Cannot connect to PingFederate with an unknown value
+	// Validate the configuration
+	if username == "" && password == "" && accessToken == "" && !hasOauthConfig {
 		resp.Diagnostics.AddError(
-			"Unable to connect to the PingFederate Server",
-			"Cannot use unknown value as password",
+			"Unable to find username and password, access_token, or OAuth configuration",
+			"username and password, access_token, or oauth configuration required values were not supplied. Either set them in the configuration or use the PINGFEDERATE_PROVIDER_* environment variables.",
+		)
+	} else if username != "" && password != "" && accessToken != "" && hasOauthConfig {
+		resp.Diagnostics.AddError(
+			"Not all values can be set",
+			"username and password, access_token, and oauth configuration cannot all be set. Only one of them can be set.",
 		)
 	} else {
-		if config.Password.IsNull() {
-			password = os.Getenv("PINGFEDERATE_PROVIDER_PASSWORD")
-		} else {
-			password = config.Password.ValueString()
+		// If user has not provided an OAuth configuration or access token, they must provide username and password
+		if !hasOauthConfig && accessToken == "" {
+			if username == "" {
+				resp.Diagnostics.AddError(
+					"Unable to find username",
+					"username cannot be an empty string. Either set it in the configuration or use the PINGFEDERATE_PROVIDER_USERNAME environment variable.",
```

Comment:
I might just combine a lot of this validation into one error message. If anything mutually exclusive is set at the same time, add one common error message that the user must choose between basic auth, access token, and oauth, along with the corresponding env vars to do so.